tests/test_core_attention.py

```diff
@@ -109,3 +110,73 @@ def test_amp_attention_sparsecs(device):
 
     expected_device = torch.float32
     assert r.dtype == expected_device
+
+
+@pytest.mark.parametrize("device", ["cuda"])
+@pytest.mark.parametrize("data_type", [torch.float16, torch.float32])
+def test_switch_blocksparse(device, data_type):
+    b, s, d = 8, 128, 32
+
+    a = torch.rand(b, s, d, device=device, dtype=data_type)
+
+    # Custom causal mask
+    m_custom = torch.triu(
+        torch.ones(s, s, device=device, dtype=a.dtype) * float("-inf"), diagonal=1
+    )
+    m_custom_bool = m_custom != float("-inf")
+    m_sparse = SparseCS(m_custom_bool, device)
+    # Mask with causal flag
+    m_att_mask = AttentionMask.make_causal(s, s, device, dtype=a.dtype)
+
+    # Check that a switch to blocksparse is only triggered by causal flag
+    with torch.cuda.amp.autocast():
+        r_custom = scaled_dot_product_attention(a, a, a, m_custom)
+        r_sparse = scaled_dot_product_attention(a, a, a, m_sparse)
+        r_att_mask = scaled_dot_product_attention(a, a, a, m_att_mask)
+
+    expected_device = torch.float32
+    assert r_sparse.dtype == expected_device
+
+    if r_custom.dtype == r_att_mask.dtype:
```

Comment:
How often is this not the case?

Reply:
When the inputs are fp16, the datatypes for normal sdp attention and blocksparse are both the same (fp16). But when the inputs are fp32, they are different where blocksparse attention returns fp32 and sdp attention always returns an fp16 tensor. I looked into this a while ago and found that it had to do with pytorch matmuls, not completely sure though ([this](https://stackoverflow.com/questions/70650236/why-does-pytorch-matmul-get-different-results-when-executed-on-cpu-and-gpu) might be relevant)

Reply:
What if you cast the results to fp16 or fp32 and compare?

Reply:
Good point, implemented!